### ISSUE TITLE
INT8 Optimization Slower Than FP32 on RTX A4000 in Windows + C++

### ISSUE THREAD

**Issue by rasikakhade** [roboflow/rf-detr#566]
I am using a custom-trained RF-DETR model exported to ONNX and trying to optimize inference on Windows 10 using ONNX Runtime with CUDA. My setup is:
GPU: NVIDIA RTX A4000, 16 GB VRAM
ONNX Runtime: 1.17 (CUDA provider)
Input size: 432×432
Precision attempts: FP32, FP16, INT8
Environment: C++ application
Problem:
1. When I run FP32, inference per image is ~17 ms.
2. When I convert to FP16, model size reduces but per-image inference increases, even using CUDA execution provider.
3. INT8 further reduces model size, but inference becomes even slower.
4. I have tried enabling EXHAUSTIVE conv search and enable_cuda_graph, but performance does not improve as expected.
5 .Batch size is currently 1; larger batch sizes are possible but I want low-latency single-image inference (~10 ms). -> while exporting onxx I tried to for dynamic batch size but it's not working, It working only for batch size 1
Attempts so far:
1. Session options: ORT_ENABLE_ALL, EXHAUSTIVE convolution search, enable_cuda_graph=1, use_tf32=1
2. Preprocessing using OpenCV to resize / normalize images
3. FP16 / INT8 export using ONNX Runtime tools
Questions / Help needed:
1. How to fully utilize FP16 Tensor Cores for single-image inference on RTX A4000?
2. Is INT8 feasible for single-image low-latency inference without calibration, or will it always be slower than FP16?
3. Are there recommended session options or memory pre-allocation techniques in C++ ONNX Runtime to achieve <10 ms per image?
4. Any Windows-specific considerations that might affect FP16/INT8 speed vs FP32?
I’d appreciate any guidance, best practices, or example C++ configurations for high-performance inference with RF-DETR on RTX A4000.
Thank you!

**Comment by isaacrob-roboflow** [comment]
We don't have one built in in this repo unfortunately, probably something work doing cc @Borda 

Ideally we'd add some kind of QAT but to be honest the post training quantization infrastructure built into PyTorch may just work for you out of the box

**Comment by isaacrob** [comment]
Ok. Should be doable, A4000 seems likely to be faster than T4 and on T4 the nano is ~3ms. 

Does that 60ms include your custom pre and post processing code? What is the raw model latency you're seeing?

**Comment by rasikakhade** [comment]
My RF-DETR segmentation inference code is currently running successfully on the CUDA backend. I attempted to set up and test the TensorRT backend to evaluate potential performance improvements; however, the inference process crashes when using TensorRT.

Below is the detailed debug output from the run. As shown, TensorRT is detected but ultimately disabled, and the session falls back to the CUDA backend. The model loads correctly, the inference session initializes without errors, and inference completes successfully on CUDA with the expected performance metrics. The crash occurs only when attempting to run inference with the TensorRT backend enabled. It is auto redirecting to CUDA.

I’m sharing the full logs below to help identify the root cause of the TensorRT inference crash.
```
?????????????????????????????????????????????????????????????
?        RFDETR Inference Engine (DEBUGGED)                ?
?        TensorRT: DISABLED  CUDA: ENABLED         ?
?????????????????????????????????????????????????????????????

[VERIFY] Checking paths...
=========================================================

ONNX Runtime library path:
  F:\RASIKA_RFDETR\onnxruntime-win-x64-gpu-cuda12-1.18.0\onnxruntime-win-x64-gpu-1.18.0\lib
  ? Path exists

Checking critical DLLs:
  ? onnxruntime.dll
  ? onnxruntime_providers_cuda.dll

ONNX Model path:
  D:\Rasika\RFDETR\RFDETR_ONNX_inferencing\inference_model_original.onnx
  ? Model exists (size: 125 MB)

Input folder:
  D:\Rasika\TESTING\test_images
  ? Folder exists

=========================================================

[PRELOAD] Loading critical DLLs...
=========================================================

CUDA Runtime:

ONNX Runtime DLLs:

Total loaded: 2 DLLs
=========================================================

[DATA] Loading images from: D:\Rasika\TESTING\test_images
[DATA] Loaded 60 images

[ONNX] Creating environment...
[ONNX] ? Environment created

[SESSION] Creating inference session...
[SESSION] Initializing inference session...
=========================================================

[PROVIDERS] Available:
  - TensorrtExecutionProvider
  - CUDAExecutionProvider
  - CPUExecutionProvider

[CUDA] Attempting to configure CUDA...
[CUDA] Appending CUDA provider...

D:\Rasika\RFDETR\RFDETR_ONNX_inferencing\RFDETR_ONNX_inferencing.exe (process 8500) exited with code -1073741819.
Press any key to close this window . . .
Press Enter to exit...
```

**Comment by isaacrob** [comment]
Ok. I have no idea why the TensorRT backend is failing in your windows environment, but that's probably the underlying reason you're not seeing the expected latencies.

**Comment by Borda** [comment]
ONNX Runtime's CUDA provider isn't optimized for transformer architectures — TensorRT consistently outperforms it. INT8 regression likely indicates uncalibrated quantization.

Consolidating quantization discussion into #504 (QAT support). For now: TensorRT FP16 on Linux for best results.

Closing — please follow #504 for updates. Feel free to reopen!
